proxy/fastcgi/proxy.go

```diff
@@ -0,0 +1,210 @@
+
+	// Create environment for CGI script
+	env, err := p.buildEnv(r, fpath)
+	if err != nil {
+		log.Printf("[WARN] failed to create fastcgi environment. %s", err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	fcgiBackend, err := Dial("tcp", p.upstream.Host)
+	if err != nil {
+		log.Printf("[WARN] failed to connect with FastCGI upstream. %s", err)
+		http.Error(w, "Bad Gateway", http.StatusBadGateway)
+		return
+	}
+	defer fcgiBackend.Close()
+
+	if err := fcgiBackend.SetReadTimeout(p.FastCGIConfig.ReadTimeout); err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if err := fcgiBackend.SetSendTimeout(p.FastCGIConfig.WriteTimeout); err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	var resp *http.Response
+
+	var contentLength int64
+	if r.ContentLength > 0 {
+		contentLength = r.ContentLength
+	} else {
+		contentLength, _ = strconv.ParseInt(r.Header.Get("Content-Length"), 10, 64)
+	}
+	switch r.Method {
+	case "HEAD":
+		resp, err = fcgiBackend.Head(env)
+	case "GET":
+		resp, err = fcgiBackend.Get(env)
+	case "OPTIONS":
+		resp, err = fcgiBackend.Options(env)
+	default:
+		resp, err = fcgiBackend.Post(env, r.Method, r.Header.Get("Content-Type"), r.Body, contentLength)
+	}
+
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	if err != nil {
+		if err, ok := err.(net.Error); ok && err.Timeout() {
+			log.Printf("[ERROR] FastCGI upstream timed out during request. %s", err)
+			http.Error(w, "Gateway Timeout", http.StatusGatewayTimeout)
+			return
+		} else if err != io.EOF {
+			log.Printf("[ERROR] failed to read response from FastCGI upstream. %s", err)
+			http.Error(w, "Bad Gateway", http.StatusBadGateway)
+			return
+		}
+	}
+
+	writeHeader(w, resp)
+
+	_, err = io.Copy(w, resp.Body)
+	if err != nil {
+		log.Printf("[ERROR] failed to write response body. %s", err)
+		http.Error(w, "Bad Gateway", http.StatusBadGateway)
+		return
+	}
+
+	if fcgiBackend.stderr.Len() != 0 {
+		log.Printf("[WARN] Error from FastCGI upstream: %s", fcgiBackend.stderr.String())
+		return
+	}
+}
+
+func writeHeader(w http.ResponseWriter, r *http.Response) {
+	for key, vals := range r.Header {
+		for _, val := range vals {
+			w.Header().Add(key, val)
+		}
+	}
+	w.WriteHeader(r.StatusCode)
+}
+
+func (p Proxy) buildEnv(r *http.Request, fpath string) (map[string]string, error) {
+	var env map[string]string
+
+	absPath := filepath.Join(p.FastCGIConfig.Root, fpath)
+
+	// Separate remote IP and port; more lenient than net.SplitHostPort
+	var ip, port string
+	if idx := strings.LastIndex(r.RemoteAddr, ":"); idx > -1 {
+		ip = r.RemoteAddr[:idx]
+		port = r.RemoteAddr[idx+1:]
+	} else {
+		ip = r.RemoteAddr
+	}
+
+	// Remove [] from IPv6 addresses
+	ip = strings.Replace(ip, "[", "", 1)
+	ip = strings.Replace(ip, "]", "", 1)
+
+	splitPos := p.splitPos(fpath)
+	if splitPos == -1 {
+		// TODO(Gufran) add appropriate error message
+		http.Error(w, "", http.StatusInternalServerError)
```

Comment:
```go
return nil, fmt.Errorf("bad address: %s", r.RemoteAddr)
```
However, if this fails then your parser is broken. So you should factor this out and write some tests for that.